### PR TITLE
[FIX] ORM: missing default values with inherits

### PR DESCRIPTION
**Pull request by danielduqma**
Description of the issue/feature this PR addresses:

Having a model A, a model B that inherits A by delegation, a model C that inherits B by delegation, and an object of model A.

An object C is created referencing existing object of model A.

Current behavior before PR:

An object of model B is created, and model A's fields are overriden with default values.

Desired behavior after PR is merged:

Object of model A keeps their original values in fields.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by danielduqma** [comment]
Hello @Feyensv, have you reviewed this PR? Thanks!

File: odoo/models.py [hunk 1]
```diff
@@ -1600,20 +1600,35 @@ def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get
...
+        for name, field in self._fields.items():
+            if name not in values and \
+                    not (self._log_access and name in MAGIC_COLUMNS):
+                if field.inherited:
+                    aux_field = field
+                    while aux_field.inherited:
+                        aux_field = aux_field.related_field
+                    if aux_field.model_name in avoid_models:
+                        continue
+                missing_defaults.add(name)
```
Comment (rco-odoo) [inline odoo/models.py:1631]: Your solution looks more complex than necessary, and I believe that `avoid_models` is not the right tool to discard inherited fields.  I suggest to build a blacklist of field names, and use it to filter out inherited fields that don't need a default:
```python
        # blacklist default values for fields inherited from a parent record
        blacklist = set()
        todo = [self]
        while todo:
            model = todo.pop()
            for parent_mname, parent_fname in model._inherits.items():
                parent_model = self.env[parent_mname]
                if parent_fname in values:
                    blacklist.update(parent_model._fields)
                else:
                    blacklist.difference_update(parent_model._fields)
                    todo.append(parent_model)

        # compute missing fields
        missing_defaults = [
            name
            for name, field in self._fields.items()
            if name not in values and not (field.inherited and name in blacklist)
        ]
```

**Comment by danielduqma** [comment]
@rco-odoo thanks for your feedback! I applied suggested changes, could you check it? I know that this PR is for version 11.0, but this bug is present in later versions.

**Comment by rco-odoo** [comment]
@danielduqma when reviewing the code I suggested to you, I realized it did not handle cases where a field is not inherited from a parent model because it is defined on the model.  For the sake of simplicity, I made my own branch, enriched the test with that specific case, and reworked your original solution to make it readable and efficient.  The result can be seen here https://github.com/odoo/odoo/pull/60898.  As you came with the first proposal, I put you as author of the commits in my branch, and mentioned myself as co-author.  Can you review my proposal, and approve it if you are okay with it?

**Comment by danielduqma** [comment]
Yes, I think your solution is better as you say and so I close this PR. Thanks for considering this!

**Pull request closed without merging**